### ISSUE TITLE
LoRA quality degradation -> xformers?

### ISSUE THREAD

**Issue by rvorias** [huggingface/diffusers#5135]
### Describe the bug

I have a forked diffusers repo that I try to keep up to date. My latest sync was from around ~3 weeks ago. This week I synced again and tried to do some quality tests with my Stable Diffusion XL LoRA inference engine. However, I noticed some really strange quality degradation.

This table corresponds to the images below:

| previous inference | current inference |
|--------|--------|
| LoRA trained w/ old diffusers | LoRA trained w/ old diffusers |
| LoRA trained w/ new diffusers | LoRA trained w/ new diffusers | 

![image](https://github.com/huggingface/diffusers/assets/21060408/e8fc85de-2f59-4caf-9c85-4b2893476b85)

There are slight differences in the old and new training scripts (vertical), but there are large and breaking differences in the inference scripts.

The old inference script is based on a custom SDXLControlnetImg2Img class. As a sanity check, I ditched my newly synced branch and installed the latest huggingface/diffusers commit, still the same behavior.

I uninstalled the latest commit and installed the latest release package (v0.21.2), still the same.

To be complete, LoRAs are loading like this:
`self.pipeline.load_lora_weights(lora_dir_path)`

Any insights would be appreciated. I suspect it's the LoRA loading refactor.

Update:
Running the same prompt more than once gives some weird results:
When I run `"a render of <mp>"` at a fixed seed I get this:
<img src="https://github.com/huggingface/diffusers/assets/21060408/76fa53ae-9867-40de-a86b-370e4f5bdb9f" width="300">

For each exact subsequent call with the exact same parameters I get:
<img src="https://github.com/huggingface/diffusers/assets/21060408/a7f2897b-551d-404f-8494-2d0afe9b094e" width="300">

Something recursive?

### Reproduction

Test a LoRA on:

(good that I know of)
https://github.com/huggingface/diffusers/commit/4909b1e3ac8446dc4ebc340cde8e5b236689ff0b

Vs new e.g.:
https://github.com/huggingface/diffusers/commit/157c9011d87e52632113024c1dc5125426971556

### Logs

Getting this after loading loras:

> diffusers/models/attention_processor.py:313: FutureWarning: `set_processor to offload LoRA` is deprecated and will be removed in version 0.26.0. In detail, removing LoRA layers via calling `set_processor` or `set_default_attn_processor` is deprecated. Please make sure to call `pipe.unload_lora_weights()` instead.
  deprecate(

### System Info

- `diffusers` version: 0.22.0.dev0
- Platform: Linux-6.2.0-32-generic-x86_64-with-glibc2.35
- Python version: 3.10.12
- PyTorch version (GPU?): 2.0.1+cu118 (True)
- Huggingface_hub version: 0.16.4
- Transformers version: 4.31.0
- Accelerate version: 0.21.0
- xFormers version: 0.0.20
- Using GPU in script?: True
- Using distributed or parallel set-up in script?: False

### Who can help?

@patrickvonplaten @sayakpaul

**Comment by sayakpaul** [comment]
We need to be able to reproduce the issue :-) Could you post some snippets for us? Make sure there are no local files in the snippet and even if there is, please attach the download instructions / code for that too.

**Comment by patrickvonplaten** [comment]
Also very curious here - it would be amazing if you could add some reproducible code snippets here @rvorias :-)

**Comment by rvorias** [comment]
I am not able to reproduce my effect with clean code on Google Colab :sob:
Will report any findings here, thanks for the feedback.

**Comment by rvorias** [comment]
Found the issue:
The issue arises when calling `pipeline.enable_xformers_memory_efficient_attention()` after `pipeline.load_lora_weights()`!

**Comment by rvorias** [comment]
@sayakpaul @patrickvonplaten FYI

**Comment by sayakpaul** [comment]
So, in the on mode, what is the problem? Is it the quality? I am unable to see anything immediately off.

**Comment by rvorias** [comment]
It's a completely different subject!

**Comment by sayakpaul** [comment]
What happens when you use PT 2.0? In PT 2.0 we use SDPA by default.

**Comment by sayakpaul** [comment]
Then it's xformers. Not very sure why xformers will act up like this.

**Comment by patrickvonplaten** [comment]
@sayakpaul can you check what's going on here?

**Comment by sayakpaul** [comment]
I am trying to disentangle the flaky behavior. Check the notebooks below:

* `nerijs/pixel-art-xl`: https://colab.research.google.com/gist/sayakpaul/ad9eb06b22ee1eac77d229def5e85f95/scratchpad.ipynb
* `ProomptEngineer/pe-balloon-diffusion-style`: https://colab.research.google.com/gist/sayakpaul/1e84d2082e36f6617e3dd774484ae456/scratchpad.ipynb
* `joachimsallstrom/aether-cloud-lora-for-sdxl`: https://colab.research.google.com/gist/sayakpaul/093b99001a12c43acabbc3842ba5cd41/scratchpad.ipynb


I've yet to observe flaky behavior. As you'd notice, I have enabled xformers too.

**Comment by sayakpaul** [comment]
Hmm, let me try by doing that.

**Comment by sayakpaul** [comment]
Didn't see anything noticeable, honestly. I did:

```
pipe.load_lora_weights(lora_repo_id, weight_name=weight_name)
pipe.enable_xformers_memory_efficient_attention()
```

And I didn't enable xformers before that.

**Comment by rvorias** [comment]
@sayakpaul https://colab.research.google.com/drive/10wiTV7bClRmCTazPsnU63qRy9fEzUbhW?usp=sharing

**Comment by sayakpaul** [comment]
I am using `pip install git+https://github.com/huggingface/diffusers`, which should install from the latest commit as well. 

Let me see what happens when I initialize the pipeline with fp16 variants directly.

**Comment by rvorias** [comment]
oh you're right, missed that

**Comment by sayakpaul** [comment]
https://colab.research.google.com/gist/sayakpaul/d17e3a34dc48bfc63ca8ea6f0102fc89/scratchpad.ipynb -- tried your set up on the other checkpoints. Still no issues.

**Comment by rvorias** [comment]
Changed to a T4, but for your checkpoints everything seems fine. Very strange. 
Then it must be my checkpoints? It cannot be the weights so it must be the state dict key names and such?
The lora weights were trained using diffusers.

**Comment by sayakpaul** [comment]
So, it seems like a device-dependent bug?

**Comment by rvorias** [comment]
Yeah seems so! I'm not sure it's the seeds. Aren't the xformer binaries compiled for different architectures?

**Comment by rvorias** [comment]
Agree, it's only noticeable on a very specific combination of device+Lora+xformers. 

Nonetheless, the effect wasn't here a few weeks ago.

Were there plans to move Lora to peft anyway? Maybe document this or have this thread as documentation.

**Comment by sayakpaul** [comment]
That is weird. 

Currently, the code will only switch to the `peft` backend when it's installed. By default, that is not the case currently. Also, `test_lora_layers_old_backend.py` has got a bunch of integration tests through which we validate the correctness. We didn't spot anything off there either. 

> Nonetheless, the effect wasn't here a few weeks ago.

Could you maybe spot the exact version that gives the expected results for you always?

**Comment by rvorias** [comment]
This is a version I was using before syncing to the newest commit:
https://colab.research.google.com/drive/10wiTV7bClRmCTazPsnU63qRy9fEzUbhW?usp=sharing

**Comment by sayakpaul** [comment]
I am also noticing that you're using a fork of `diffusers` and not the actual one.

**Comment by rvorias** [comment]
They should be interchangeable, will rerun!

**Comment by patrickvonplaten** [comment]
The bug is quite visible here: https://colab.research.google.com/drive/1_2QHvFjGNdSlZfLRs0csmMt2HXY11NJj?usp=sharing

The question is not whether LoRA works or not, the question is whether LoRA + xformers gives different results to LoRA without xformers

**Comment by patrickvonplaten** [comment]
Found the bug, I introduced it :sweat_smile: - fixing it now!
